### PR TITLE
Catch an underflow issue in DataflowPipelineJob's backoff code

### PR DESCRIPTION
**Pull request by dhalperi**

### CODE REVIEW

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/DataflowPipelineJob.java [hunk 1]
```diff
@@ -255,19 +255,26 @@ State waitToFinish(
...
-              MESSAGES_BACKOFF_FACTORY.withMaxCumulativeBackoff(duration.minus(consumed)).backoff();
-        }
-        // Check if the job is done.
-        if (state.isTerminal()) {
-          return state;
+          Duration remaining = duration.minus(consumed);
+          if (remaining.isLongerThan(Duration.ZERO)) {
+            backoff = MESSAGES_BACKOFF_FACTORY.withMaxCumulativeBackoff(remaining).backoff();
+          } else {
+            backoff = BackOff.STOP_BACKOFF;
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/DataflowPipelineJob.java]:
```
// If there is no time remaining, don't bother backing off.
```

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/DataflowPipelineJob.java [hunk 2]
```diff
@@ -255,19 +255,26 @@ State waitToFinish(
...
           Duration consumed = Duration.millis((nanosConsumed + 999999) / 1000000);
-          backoff =
-              MESSAGES_BACKOFF_FACTORY.withMaxCumulativeBackoff(duration.minus(consumed)).backoff();
-        }
-        // Check if the job is done.
-        if (state.isTerminal()) {
-          return state;
+          Duration remaining = duration.minus(consumed);
+          if (remaining.isLongerThan(Duration.ZERO)) {
+            backoff = MESSAGES_BACKOFF_FACTORY.withMaxCumulativeBackoff(remaining).backoff();
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/DataflowPipelineJob.java]: `withMaxCumulativeBackoff(...)` and `withMaxBackoff(...)` should have a `Preconditions.checkArgument` that throws an appropriate error when misspecified.

**Comment by bjchambers** [comment]
LGTM -- not sure why travis is sad

**Comment by dhalperi** [comment]
Travis Mac builds are dead, but the 3 linux builds passed. Merging.

**Pull request merged**